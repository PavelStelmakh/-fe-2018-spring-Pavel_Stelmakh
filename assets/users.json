[
  {
    "id": 1,
    "name": "Pasha",
    "age": "19",
    "password": "1111",
    "dateOfBirth": "2011-10-05T00:00:00.000Z",
    "dateOfFirstLogin": "2011-10-05T14:48:00.000Z",
    "dateOfNextNotif": "2011-10-05T14:48:00.000Z",
    "information": "meaw"
  },
  {
    "id": 2,
    "name": "Sasha",
    "age": "24",
    "password": "2222",
    "dateOfBirth": "2011-10-05T00:00:00.000Z",
    "dateOfFirstLogin": "2011-10-05T14:48:00.000Z",
    "dateOfNextNotif": "2011-10-05T14:48:00.000Z",
    "information": "meaw"
  },
  {
    "id": 3,
    "name": "Masha",
    "age": "20",
    "password": "3333",
    "dateOfBirth": "2011-10-05T00:00:00.000Z",
    "dateOfFirstLogin": "2011-10-05T14:48:00.000Z",
    "dateOfNextNotif": "2011-10-05T14:48:00.000Z",
    "information": "meaw"
  }
]