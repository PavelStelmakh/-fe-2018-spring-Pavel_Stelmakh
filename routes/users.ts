import { Router, Response, Request, NextFunction } from 'express';
const router = Router();
import { User } from '../models/User';
const loginRoute = require('./login');
const users = require('../models/Users');
import { getUserByToken } from '../shared/getUserByToken';

router.use('/login', loginRoute);

router.use('*', (req: Request, res: Response, next: NextFunction) => {
    const data = getUserByToken(req);

    if (data.user) {
        next();
    } else {
        res.status(401);
        res.end();
    }
});

router.get('/users', (_req: Request, res: Response) => {
    res.status(200).send(users.users());
});

router.get('/users/search/:name', (req: Request, res: Response) => {
    try {
        const name: number = req.params['name'];
        const searchUsers: User[] = users.search(name);
        setTimeout(() => {
            if (searchUsers.length > 0) {
                res.status(200).send(searchUsers);
            } else {
                res.status(400);
                res.end();
            }
        }, 1000);
    }
    catch (Error) {
        res.status(418);
        res.end();
    }
});

router.get('/users/:id', (req: Request, res: Response) => {
    try {
        const id: number = +req.params['id'];
        const user: User | undefined = users.user(id);
        if (user) {
            setTimeout(() => res.status(200).send(user), 3000);
        } else {
            res.status(400);
            res.end();
        }
    }
    catch (Error) {
        res.status(418);
        res.end();
    }
});

router.post('/users/add', (req: Request, res: Response) => {
    try {
        const user: User = users.add(req.body);
        res.status(200).send(user);
    }
    catch (Error) {
        res.status(418);
    }
    res.end();
});

router.put('/users/:id', (req: Request, res: Response) => {
    try {
        const id: number = +req.params['id'];
        const success: boolean = users.update(id, req.body);
        if (success) {
            res.status(200);
        } else {
            res.status(418);
        }
    }
    catch (Error) {
        res.status(500);
    }
    res.end();
});

router.delete('/users/:id', (req: Request, res: Response) => {
    try {
        const id: number = +req.params['id'];
        const success: boolean = users.delete(id);
        if (success) {
            res.status(200);
        } else {
            res.status(418);
        }
    }
    catch (Error) {
        res.status(500);
    }
    res.end();
});

module.exports = router;