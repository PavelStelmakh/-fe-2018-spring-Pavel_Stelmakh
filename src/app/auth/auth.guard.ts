import { Injectable } from '@angular/core';
import { CanLoad, ActivatedRouteSnapshot, RouterStateSnapshot, Router, Route, CanActivate } from '@angular/router';
import { Observable, of } from 'rxjs';
import { mergeMap, take, catchError } from 'rxjs/operators';
import { Store } from '@ngrx/store';
import * as profileAction from '../actions/profile.action';
import * as profileReducer from '../reducers/profile.reduser';
import { AuthService } from './auth.service';

@Injectable({
  providedIn: 'root'
})
export class AuthGuard implements CanLoad, CanActivate {
  constructor(
    private router: Router,
    private authService: AuthService,
    private store: Store<profileReducer.State>
    ) {}

    canLoad(route: Route): Observable<boolean> | Promise<boolean> | boolean {
      const {path} = route;
      return this.authService.checkLogin().pipe(
        take(1),
        mergeMap(result => {
          if (result.body['role'] === 'admin') {
            this.store.dispatch(new profileAction.SetIdAction(+result.body['id']));
            if (path === 'admin') {
              return of(true);
            } else {
              this.router.navigate(['/admin']);
              return of(false);
            }
          } else {
            if (path === 'user') {
              return of(true);
            } else {
              this.router.navigate(['/user']);
              return of(false);
            }
          }
        }),
        catchError(error => {
          this.router.navigate(['/login']);
          return of(false);
        })
      );
  }

  canActivate(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<boolean> | Promise<boolean> | boolean {
    return this.authService.checkLogin().pipe(
    take(1),
    mergeMap(result => {
      this.store.dispatch(new profileAction.SetIdAction(+result.body['id']));
      switch (state.url) {
        case '/login':
        case '/login/recovery':
          this.router.navigate([`/${result.body['role']}`]);
          return of(false);
        default: return of(true);
      }   
    }),
    catchError(error => {
      switch (state.url) {
        case '/login':
        case '/login/recovery': 
          return of(true);
        default: 
          this.router.navigate(['/login']);
          return of(false);
        }
      })
    );
  }

}
