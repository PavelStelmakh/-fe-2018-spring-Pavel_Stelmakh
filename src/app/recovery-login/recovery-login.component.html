<form [formGroup]="recoveryForm" (ngSubmit)="onSubmit()">
  <div *ngIf="wrongData">{{'login.wrongData' | translate}}</div>
  <div *ngIf="login.dirty && login.errors?.checkExistName">{{ 'login.notFound' | translate }}</div>
  <div class="loading" *ngIf="login.dirty && login.status === 'PENDING'">loading...</div>
  <input type="text" placeholder="login" formControlName="login">
  <div *ngIf="birthday.dirty && birthday.errors?.date">{{ 'user-details.user-form.messages.dateOfBirth' | translate }}</div>
  <input type="text" placeholder="date of birthday (YYYY/MM/DD)" formControlName="birthday">
  <input type="password" placeholder="new password" formControlName="password">
  <input type="submit" [disabled]="recoveryForm.pristine || recoveryForm.status !== 'VALID'" value="{{'login.recovery' | translate}}">
  <a routerLink="/login">{{'login.back' | translate}}</a>
</form>
