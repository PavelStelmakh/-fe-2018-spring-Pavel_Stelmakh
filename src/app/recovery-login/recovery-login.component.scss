@import '../shared/driveAnimation';

@include animation("recovery", 0, 20vh, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.7));
@include animation("message", -40px, 0, rgba(250, 86, 85, 0), rgba(250, 86, 85, 1));

:host {    
    text-align: center;
    display: block;
    width: 500px;
    background-color: rgba(0, 0, 0, 0.7);
    position: absolute;
    top: 20vh;
    left: calc(50vw - 250px);
    box-shadow: 0 0 10px 3px rgba(0,0,0,0.5);
    animation: recoveryAnimation 1s;

    form {
        padding: 10px 0;
    }

}

.recovery {
    margin-top: 10px;
}

a {    
    color: #ccc;
    margin-left: 10px;
}

.messages {
    position: absolute;
    top: -13vh;
    height: 40px;
    line-height: 40px;
    width: 500px;
    color: #ccc;

    div {
        position: relative;
        top: 0;
        background-color: rgba(250, 86, 85, 1);
        box-shadow: 0 0 10px 3px rgba(0, 0, 0, 0.5);
        animation:  messageAnimation 1s;
    }

    .loading {
        background-color: rgb(250, 86, 85);
    }

}