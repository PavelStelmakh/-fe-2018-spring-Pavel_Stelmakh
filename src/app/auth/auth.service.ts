import { Injectable, Inject } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { BASE_URL } from '../url.service';
import { SignIn } from 'models/SignIn';
import { Recovery } from 'models/Recovery';
import { tap } from 'rxjs/operators';
import { UsersService } from '../users.service';

@Injectable({
  providedIn: 'root'
})
export class AuthService {

  constructor(
    private http: HttpClient, 
    @Inject(BASE_URL) private url: string,
    private user: UsersService
    ) { }

  signIn(userData: SignIn) {
    return this.http.post<SignIn>(`${this.url}/login`, userData, {
      headers: new HttpHeaders({
        'Content-Type':  'application/json'
      }),
      observe: 'response',
      withCredentials: true
    });
  }

  recovery(userData: Recovery) {
    return this.http.post<SignIn>(`${this.url}/login/recovery`, userData, {
      headers: new HttpHeaders({
        'Content-Type':  'application/json'
      }),
      observe: 'response'
    });
  }

  checkLogin() {
    return this.http.get(`${this.url}/login`, {
      observe: 'response',
      withCredentials: true
    })
    .pipe(
      tap(result => {
        if (result.status === 200) {
          this.user.setId(+result.body['id']);
        }
      })
    );
  }

  logout() {
    return this.http.get(`${this.url}/login/logout`, {
      observe: 'response',
      withCredentials: true
    });
  }

}
