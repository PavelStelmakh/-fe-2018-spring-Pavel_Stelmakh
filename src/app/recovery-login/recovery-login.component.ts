import { Component, OnInit } from '@angular/core';
import { FormBuilder, FormGroup, Validators } from '@angular/forms';
import { AuthService } from '../auth/auth.service';
import { Router } from '@angular/router';
import { Recovery } from 'models/Recovery';
import { dateValidator, checkExistNameValidator } from '../validators';
import { UsersService } from '../users.service';

@Component({
  selector: 'app-recovery-login',
  templateUrl: './recovery-login.component.html',
  styleUrls: ['./recovery-login.component.scss']
})
export class RecoveryLoginComponent implements OnInit {
  recoveryForm: FormGroup;
  wrongData: boolean;

  constructor(
    private builder: FormBuilder, 
    private auth: AuthService, 
    private router: Router, 
    private usersService: UsersService
    ) {}

  ngOnInit() {
    this.wrongData = false;
    this.recoveryForm = this.builder.group({
      login: ['', Validators.required, checkExistNameValidator(this.usersService, true)],
      birthday: ['', [Validators.required, dateValidator('YYYY/MM/DD')]],
      password: ['', Validators.required]
    });
  }

  get login() {
    return this.recoveryForm.get('login');
  }

  get birthday() {
    return this.recoveryForm.get('birthday');
  }

  get password() {
    return this.recoveryForm.get('password');
  }

  onSubmit() {
    const data: Recovery = {
      login: this.login.value, 
      birthday: this.birthday.value, 
      password: this.password.value
    };
    this.auth.recovery(data).subscribe(result => {
      if (result.status === 200) {
        this.router.navigate(['/login']);
      }
    },
    error => {
      this.recoveryForm.reset();
      this.wrongData = true;
    });
    
  }

}
