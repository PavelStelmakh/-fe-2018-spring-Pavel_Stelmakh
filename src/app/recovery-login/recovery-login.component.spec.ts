import { TestBed, async, ComponentFixture } from '@angular/core/testing';
import { RouterTestingModule } from "@angular/router/testing";
import { HttpClientModule } from '@angular/common/http';
import { HttpClientTestingModule } from '@angular/common/http/testing';
import { ReactiveFormsModule } from '@angular/forms';
import { NO_ERRORS_SCHEMA } from '@angular/core';
import { RecoveryLoginComponent } from './recovery-login.component';
import { UsersService } from '../users.service';
import { AuthService } from '../auth/auth.service';
import { BASE_URL } from '../url.service';
import { Observable, of } from 'rxjs';
import { TranslateModule, TranslateLoader, TranslateService } from '@ngx-translate/core';

const translations = {
  "login": {
    "wrongData": "wrong data",
    "notFound": "such login is not found",
    "back": "back",
    "recovery": "recovery",
    "user-details": {
      "user-form": {
        "messages": {
          "dateOfBirth": "date should have format YYYY/MM/DD"
        }
      }
    }
  }
};

class FakeLoader implements TranslateLoader {
  getTranslation(lang: string): Observable<{}> {
    return of(translations);
  }
}

describe('RecoveryLoginComponent', () => {
  let fixture: ComponentFixture<RecoveryLoginComponent>;
  let translate: TranslateService;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [
        RouterTestingModule,
        HttpClientModule,
        HttpClientTestingModule,
        ReactiveFormsModule,
        TranslateModule.forRoot({
          loader: {provide: TranslateLoader, useClass: FakeLoader}
        })
      ],
      providers: [
        {provide: BASE_URL, useValue: 'http://localhost:3000'},
        UsersService,
        AuthService
      ],
      declarations: [
        RecoveryLoginComponent
      ],
      schemas: [
       NO_ERRORS_SCHEMA
      ]
    }).compileComponents();

  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(RecoveryLoginComponent);
    translate = TestBed.get(TranslateService);
    translate.use('en');
  });

  it('renders markup to snapshot', () => {
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });

});
