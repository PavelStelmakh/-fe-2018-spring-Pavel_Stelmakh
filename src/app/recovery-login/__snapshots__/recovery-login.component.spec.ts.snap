// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecoveryLoginComponent renders markup to snapshot 1`] = `
<app-recovery-login
  auth={[Function AuthService]}
  builder={[Function FormBuilder]}
  recoveryForm={[Function FormGroup]}
  router={[Function Router]}
  usersService={[Function UsersService]}
  wrongData="false"
>
  <form
    class="ng-untouched ng-pristine ng-invalid"
    ng-reflect-form="[object Object]"
    novalidate=""
  >
    <div
      class="messages"
    >
      
      
      
      
    </div>
    <input
      class="ng-untouched ng-pristine ng-invalid"
      formcontrolname="login"
      ng-reflect-name="login"
      placeholder="login"
      type="text"
    />
    <input
      class="ng-untouched ng-pristine ng-invalid"
      formcontrolname="birthday"
      ng-reflect-name="birthday"
      placeholder="date of birthday (YYYY/MM/DD)"
      type="text"
    />
    <input
      class="ng-untouched ng-pristine ng-invalid"
      formcontrolname="password"
      ng-reflect-name="password"
      placeholder="new password"
      type="password"
    />
    <div
      class="recovery"
    >
      <input
        disabled=""
        type="submit"
        value="recovery"
      />
      <a
        href="/login"
        ng-reflect-router-link="/login"
        routerlink="/login"
      >
        back
      </a>
    </div>
  </form>
</app-recovery-login>
`;
